relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/server/CogRPCService.kt

```diff
@@ -0,0 +1,97 @@
+import tech.relaycorp.relaynet.cogrpc.CargoRelayGrpc
+import java.util.logging.Level
+import java.util.logging.Logger
+
+class CogRPCService(
+    private val coroutineScope: CoroutineScope,
+    private val connectionService: CargoRelayServer.ConnectionService
+) : CargoRelayGrpc.CargoRelayImplBase() {
+
+    override fun deliverCargo(responseObserver: StreamObserver<CargoDeliveryAck>) =
+        object : StreamObserver<CargoDelivery> {
+            override fun onNext(cargoDelivery: CargoDelivery) {
+                coroutineScope.launch {
+                    logger.info("deliverCargo next")
+                    connectionService.deliverCargo(cargoDelivery.toMessageReceived())
+                    logger.info("deliverCargo next ack")
+                    responseObserver.onNext(cargoDelivery.toAck())
+                }
+            }
+
+            override fun onError(throwable: Throwable) {
+                logger.log(Level.WARNING, "deliverCargo error", throwable)
+            }
+
+            override fun onCompleted() {
+                logger.info("deliverCargo complete")
+                coroutineScope.launch {
+                    responseObserver.onCompleted()
+                }
+            }
+        }
+
+    override fun collectCargo(responseObserver: StreamObserver<CargoDelivery>): StreamObserver<CargoDeliveryAck> {
+        val cca = Authorization.getCCA()
+        coroutineScope.launch {
+            try {
+                val messageReceived = CargoRelay.MessageReceived(cca.inputStream())
+                val deliveries = connectionService.collectCargo(messageReceived)
+                deliveries.forEach { messageDelivery ->
+                    logger.info("collectCargo delivering ${messageDelivery.localId}")
+                    responseObserver.onNext(messageDelivery.toCargoDelivery())
+                }
+            } catch (exception: Exception) {
+                logger.log(Level.SEVERE, "collectCargo error", exception)
+            }
+            logger.info("collectCargo completed")
+            responseObserver.onCompleted()
+        }
+
+        return object : StreamObserver<CargoDeliveryAck> {
+            override fun onNext(ack: CargoDeliveryAck) {
+                logger.info("collectCargo ack next")
+                coroutineScope.launch {
+                    connectionService.processCargoCollectionAck(ack.toMessageDeliveryAck())
+                }
+            }
+
+            override fun onError(throwable: Throwable) {
+                logger.log(Level.WARNING, "collectCargo ack error", throwable)
+            }
+
+            override fun onCompleted() {
+                logger.info("collectCargo ack complete")
+            }
+        }
+    }
+
+    internal fun CargoDelivery.toMessageReceived() =
+        CargoRelay.MessageReceived(data = cargo.newInput())
+
+    private fun CargoRelay.MessageDelivery.toCargoDelivery() =
+        CargoDelivery.newBuilder()
+            .setId(localId)
```

Comment:
N00b question alert :D : Where's `localId` coming from? And more importantly, is it a UUID4 or is it based on `SecureRandom`?

Reply:
It's a simple one derived from the StoredMessage, so can retrieve it back: https://github.com/relaycorp/relaynet-courier-android/blob/9a18e9c1245d27a440ba464b55bb73b6cd3b56e0/app/src/main/java/tech/relaycorp/courier/domain/client/UniqueMessageId.kt